src/pykeen/nn/vision.py

```diff
@@ -0,0 +1,297 @@
+            Should not include `max_id` or `shape`.
+
+        :raises ValueError:
+            if `max_id` is provided and does not match the number of images
+        """
+        _ensure_vision(self, models)
+        self.images = VisionDataset(images=images, transforms=transforms)
+
+        if isinstance(encoder, str):
+            cls = getattr(models, encoder)
+            encoder = cls(encoder_kwargs or {})
+
+        pool = functools.partial(torch.mean, dim=(-1, -2))
+
+        encoder = models.feature_extraction.create_feature_extractor(
+            model=encoder, return_nodes={layer_name: "feature"}
+        )
+
+        # infer shape
+        with torch.inference_mode():
+            encoder.eval()
+            shape = self._encode(images=self.images[0].unsqueeze(dim=0), encoder=encoder, pool=pool).shape[1:]
+
+        if max_id is not None and len(images) != max_id:
+            raise ValueError(
+                f"Inconsistent max_id={max_id} and len(images)={len(images)}. In case there are not images for all "
+                f"IDs, you may consider using BackfillRepresentation.",
+            )
+        super().__init__(max_id=max_id, shape=shape, **kwargs)
+
+        self.encoder = encoder
+        self.pool = pool
+        self.batch_size = batch_size or self.max_id
+        self.encoder.train(trainable)
+        self.encoder.requires_grad_(trainable)
+        self.trainable = trainable
+
+    @staticmethod
+    def _encode(
+        images: torch.FloatTensor, encoder: torch.nn.Module, pool: Callable[[torch.FloatTensor], torch.FloatTensor]
+    ) -> torch.FloatTensor:
+        """
+        Encode images with the given encoder and pooling methods.
+
+        :param images: shape: (batch_size, num_channels, height, width)
+            a batch of images
+        :param encoder:
+            the encoder, returning a dictionary with key "features"
+        :param pool:
+            the pooling method to use
+        :return: shape: (batch_size, dim)
+            the encoded representations.
+        """
+        return pool(encoder(images)["feature"])
+
+    # docstr-coverage: inherited
+    def _plain_forward(self, indices: Optional[torch.LongTensor] = None) -> torch.FloatTensor:  # noqa: D102
+        dataset = self.images
+        if indices is not None:
+            dataset = torch.utils.data.Subset(dataset=dataset, indices=indices)
+        data_loader = torch.utils.data.DataLoader(dataset, batch_size=self.batch_size)
+        with torch.inference_mode(mode=not self.trainable):
+            return torch.cat(
+                [self._encode(images=images, encoder=self.encoder, pool=self.pool) for images in data_loader], dim=-1
+            )
+
+
+class WikidataVisualRepresentation(BackfillRepresentation):
+    """
+    Visual representations obtained from Wikidata and encoded with a vision encoder.
+
+    If no image could be found for a certain Wikidata ID, a plain (trainable) embedding will be used instead.
```

Comment:
hellaaaaaa. sort of bizarre the way that the class inheritance works but good enough